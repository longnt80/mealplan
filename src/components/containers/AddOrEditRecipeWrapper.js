import React, { Component } from 'react';
import PropTypes from 'prop-types';
import firebase from '../../config/fbConfig';

import AddOrEditRecipe from '../AddOrEditRecipe';

const generateNewRecipe = () => {
  const time = Date.now();

  return {
    recipeName: '',
    ingredients: [
      {
        name: "",
        amount: 0,
        unit: "",
        id: `ingr-${time}`
      }
    ],
    direction: "",
  };
}

const STATUS = {
  ADD: 'add',
  EDIT: 'edit'
}

class AddOrEditRecipeWrapper extends Component {
  static propTypes = {
    location: PropTypes.shape({
      state: PropTypes.shape({
        recipe: PropTypes.shape({
          recipeName: PropTypes.string,
          direction: PropTypes.string,
          ingredients: PropTypes.arrayOf(PropTypes.shape({
            amount: PropTypes.number,
            id: PropTypes.string,
            name: PropTypes.string,
            unit: PropTypes.string
          }))
        })
      })
    }),

  }

  constructor(props) {
    super(props)
    const { state } = this.props.location;
    const recipeIsExisted = state && state.recipe;
    const recipe = recipeIsExisted ? state.recipe : generateNewRecipe();
    const status = recipeIsExisted ? STATUS.EDIT : STATUS.ADD;
    this.state = {
      status,
      recipe: {...recipe}
    }
  }

  addIngredientField = currentValues => e => {
    const time = Date.now();

    const emptyIngredientFields = {
      name: "",
      amount: 0,
      unit: "",
      id: `ingr-${time}`,
    };

    this.setState({
      recipe: {
        ...currentValues,
        ingredients: [
          ...currentValues.ingredients,
          emptyIngredientFields,
        ]
      }
    });
  }

  deleteIngredientField = (id, currentValues) => {
    console.log("delete ingr field: " + id);

    const newIngredientList = currentValues.ingredients.filter(ingredient => ingredient.id !== id);

    this.setState({
      recipe: {
        ...currentValues,
        ingredients: [
          ...newIngredientList
        ]
      }
    });
  }

  addRecipe = recipe => {
    const db = firebase.firestore();

    return db.collection('recipes').add({
      ...recipe
    });
  }

  updateRecipe = recipe => {
    const db = firebase.firestore();
    const { id } = this.props.match.params;

    return db.collection('recipes').doc(id).set(recipe);
  }

  handleFormSubmit = (recipe, status) => {
    const { history } = this.props;
    const { id } = this.props.match.params;

    // Remove empty ingredient fields
    const filteredIngredientsList = recipe.ingredients.filter(ingredient => ingredient.name !== "");
    const validRecipe = {
      ...recipe,
      ingredients: [
        ...filteredIngredientsList
      ]
    }

    if(status === STATUS.ADD) {
      this.addRecipe(validRecipe)
      .then(() => console.log("Added recipe"));
    } else if (status === STATUS.EDIT) {
      this.updateRecipe(validRecipe)
      .then(() => console.log("Updated recipe"));
    }

    history.push(`/recipe/view/${id}`, { recipe: { ...validRecipe } });
  }

  render() {
    console.log(this.state.status);
    return (
      <AddOrEditRecipe
        currentStatus={this.state.status}
        initialFields={this.state.recipe}
        addIngredientField={this.addIngredientField}
        deleteIngredientField={this.deleteIngredientField}
        handleFormSubmit={this.handleFormSubmit}
      />
    );
  }
}

export default AddOrEditRecipeWrapper;
