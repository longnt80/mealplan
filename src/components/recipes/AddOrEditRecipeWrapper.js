import React, { Component } from 'react';
import PropTypes, { any } from 'prop-types';
import firebase from '../../config/fbConfig';

import { DB_RECIPES_COLLECTION } from '../../common/constants';

import AddOrEditRecipe from './AddOrEditRecipe';

const db = firebase.firestore();

const generateNewRecipe = () => {
  const time = Date.now();

  return {
    recipeName: '',
    ingredients: [
      {
        name: "",
        amount: 0,
        unit: "",
        id: `ingr-${time}`
      }
    ],
    direction: "",
  };
}

export const STATUS = {
  ADD: 'add',
  EDIT: 'edit'
}

class AddOrEditRecipeWrapper extends Component {
  static propTypes = {
    location: PropTypes.shape({
      state: PropTypes.shape({
        recipe: PropTypes.shape({
          recipeName: PropTypes.string,
          direction: PropTypes.string,
          ingredients: PropTypes.arrayOf(PropTypes.shape({
            amount: PropTypes.number,
            id: PropTypes.string,
            name: PropTypes.string,
            unit: PropTypes.string
          }))
        })
      })
    }),
    match: PropTypes.shape({
      params: PropTypes.shape({
        id: PropTypes.string
      })
    }),
    history: PropTypes.objectOf(any),
  }

  static defaultProps = {
    location: {},
    match: {},
    history: {},
  }

  constructor(props) {
    super(props)
    const { location } = this.props;
    const recipeIsExisted = location.state && location.state.recipe;
    const recipe = recipeIsExisted ? location.state.recipe : generateNewRecipe();
    const status = recipeIsExisted ? STATUS.EDIT : STATUS.ADD;
    this.state = {
      status,
      recipe: {...recipe}
    }
  }

  addIngredientField = currentValues => {
    const time = Date.now();

    const emptyIngredientFields = {
      name: "",
      amount: 0,
      unit: "",
      id: `ingr-${time}`,
    };

    this.setState({
      recipe: {
        ...currentValues,
        ingredients: [
          ...currentValues.ingredients,
          emptyIngredientFields,
        ]
      }
    });
  }

  deleteIngredientField = (id, currentValues) => {
    const newIngredientList = currentValues.ingredients.filter(ingredient => ingredient.id !== id);

    this.setState({
      recipe: {
        ...currentValues,
        ingredients: [
          ...newIngredientList
        ]
      }
    });
  }

  addRecipe = recipe => {
    return db.collection(DB_RECIPES_COLLECTION).add({
      ...recipe
    });
  }

  updateRecipe = recipe => {
    const { match } = this.props;

    return db.collection(DB_RECIPES_COLLECTION).doc(match.params.id).set(recipe, { merge: true });
  }

  deleteRecipe = () => {
    const { history } = this.props;
    const { match } = this.props;

    db.collection(DB_RECIPES_COLLECTION).doc(match.params.id).delete()
      .then(() => {
        console.log("Deleted");
        history.push('/recipes')
      });
  }

  handleFormSubmit = (recipe, status) => {
    const { history } = this.props;
    const { match } = this.props;

    // Remove empty ingredient fields
    const filteredIngredientsList = recipe.ingredients.filter(ingredient => ingredient.name !== "");
    const validRecipe = {
      ...recipe,
      ingredients: [
        ...filteredIngredientsList
      ]
    }

    if(status === STATUS.ADD) {
      this.addRecipe(validRecipe)
        .then(() => console.log("Added recipe"));
    } else if (status === STATUS.EDIT) {
      this.updateRecipe(validRecipe)
        .then(() => console.log("Updated recipe"));
    }

    history.push(`/recipe/view/${match.params.id}`, { recipe: { ...validRecipe } });
  }

  render() {
    const { status, recipe } = this.state;

    return (
      <AddOrEditRecipe
        currentStatus={status}
        initialFields={recipe}
        deleteRecipe={status === STATUS.EDIT ? this.deleteRecipe : null}
        addIngredientField={this.addIngredientField}
        deleteIngredientField={this.deleteIngredientField}
        handleFormSubmit={this.handleFormSubmit}
      />
    );
  }
}

export default AddOrEditRecipeWrapper;
