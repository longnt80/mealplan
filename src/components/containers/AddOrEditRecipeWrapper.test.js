import React from 'react';
import AddOrEditRecipeWrapper from './AddOrEditRecipeWrapper';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
  const component = renderer.create(
    <AddOrEditRecipeWrapper
      location={
        {
          state: {
            recipe: {
              description: "You need to make something",
              recipeName: "A hamburger",
              ingredients: [
                {
                  name: "Long",
                  id: "adasdasdasdasdasdasd",
                  amount: 1,
                  unit: "ml"
                },
                {
                  name: "Phuong",
                  id: "adasdasd12345678",
                  amount: 1,
                  unit: "kg"
                }
              ]
            }
          }
        }
      }
    />
  ).toJSON();

  expect(component).toMatchSnapshot();
});
