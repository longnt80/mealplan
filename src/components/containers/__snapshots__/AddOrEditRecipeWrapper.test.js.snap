// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="MuiPaper-root-4 MuiPaper-elevation2-8 MuiPaper-rounded-5 MyForm-paper-1"
  >
    <h2
      className="MuiTypography-root-31 MuiTypography-h5-47"
    >
      Recipe's name:
    </h2>
    <div
      className="MuiFormControl-root-67"
    >
      <label
        className="MuiFormLabel-root-82 MuiFormLabel-filled-86 MuiInputLabel-root-71 MuiInputLabel-formControl-76 MuiInputLabel-animated-79 MuiInputLabel-shrink-78"
        data-shrink={true}
      >
        Name
      </label>
      <div
        className="MuiInputBase-root-102 MuiInput-root-89 MuiInput-underline-93 MuiInputBase-formControl-103 MuiInput-formControl-90"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input-112 MuiInput-input-97"
          disabled={false}
          name="recipeName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value="A hamburger"
        />
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root-4 MuiPaper-elevation2-8 MuiPaper-rounded-5 MyForm-paper-1"
  >
    <h2
      className="MuiTypography-root-31 MuiTypography-h5-47"
    >
      Ingredients:
    </h2>
    <div
      className="MuiGrid-container-119 MuiGrid-spacing-xs-24-143 MuiGrid-justify-xs-space-between-138"
    >
      <div
        className="MuiGrid-item-120 MuiGrid-grid-xs-12-159 MuiGrid-grid-sm-6-167"
      >
        <div
          className="MuiFormControl-root-67 MuiFormControl-fullWidth-70"
        >
          <label
            className="MuiFormLabel-root-82 MuiFormLabel-filled-86 MuiInputLabel-root-71 MuiInputLabel-formControl-76 MuiInputLabel-animated-79 MuiInputLabel-shrink-78"
            data-shrink={true}
          >
            Ingredient
          </label>
          <div
            className="MuiInputBase-root-102 MuiInput-root-89 MuiInput-underline-93 MuiInputBase-fullWidth-111 MuiInput-fullWidth-96 MuiInputBase-formControl-103 MuiInput-formControl-90"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-112 MuiInput-input-97"
              disabled={false}
              name="ingredients[0].name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="Long"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-item-120 MuiGrid-grid-xs-12-159 MuiGrid-grid-sm-2-163"
      >
        <div
          className="MuiGrid-container-119 MuiGrid-justify-xs-center-136"
        >
          <button
            className="MuiButtonBase-root-222 MuiIconButton-root-216"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="button"
          >
            <span
              className="MuiIconButton-label-221"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-225"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-275"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-container-119 MuiGrid-spacing-xs-24-143 MuiGrid-justify-xs-space-between-138"
    >
      <div
        className="MuiGrid-item-120 MuiGrid-grid-xs-12-159 MuiGrid-grid-sm-6-167"
      >
        <div
          className="MuiFormControl-root-67 MuiFormControl-fullWidth-70"
        >
          <label
            className="MuiFormLabel-root-82 MuiFormLabel-filled-86 MuiInputLabel-root-71 MuiInputLabel-formControl-76 MuiInputLabel-animated-79 MuiInputLabel-shrink-78"
            data-shrink={true}
          >
            Ingredient
          </label>
          <div
            className="MuiInputBase-root-102 MuiInput-root-89 MuiInput-underline-93 MuiInputBase-fullWidth-111 MuiInput-fullWidth-96 MuiInputBase-formControl-103 MuiInput-formControl-90"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-112 MuiInput-input-97"
              disabled={false}
              name="ingredients[1].name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value="Phuong"
            />
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-item-120 MuiGrid-grid-xs-12-159 MuiGrid-grid-sm-2-163"
      >
        <div
          className="MuiGrid-container-119 MuiGrid-justify-xs-center-136"
        >
          <button
            className="MuiButtonBase-root-222 MuiIconButton-root-216"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="-1"
            type="button"
          >
            <span
              className="MuiIconButton-label-221"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-225"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
              </svg>
            </span>
            <span
              className="MuiTouchRipple-root-275"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      className="MyForm-buttonContainer-3"
    >
      <button
        className="MuiButtonBase-root-222 MuiButton-root-234 MuiButton-contained-245 MuiButton-raised-248 MuiButton-sizeSmall-257"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiButton-label-235"
        >
          More ingredient
        </span>
        <span
          className="MuiTouchRipple-root-275"
        />
      </button>
    </div>
  </div>
  <div
    className="MuiPaper-root-4 MuiPaper-elevation2-8 MuiPaper-rounded-5 MyForm-paper-1"
  >
    <h2
      className="MuiTypography-root-31 MuiTypography-h5-47"
    >
      Direction:
    </h2>
    <div
      className="MuiFormControl-root-67 MuiFormControl-fullWidth-70"
    >
      <label
        className="MuiFormLabel-root-82 MuiFormLabel-filled-86 MuiInputLabel-root-71 MuiInputLabel-formControl-76 MuiInputLabel-animated-79 MuiInputLabel-shrink-78 MuiInputLabel-outlined-81"
        data-shrink={true}
      >
        Direction
      </label>
      <div
        className="MuiInputBase-root-102 MuiOutlinedInput-root-260 MuiInputBase-fullWidth-111 MuiInputBase-formControl-103 MuiInputBase-multiline-110 MuiOutlinedInput-multiline-266"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-273 MuiOutlinedInput-notchedOutline-267"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-274"
            style={
              Object {
                "width": 0,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <textarea
          aria-invalid={false}
          className="MuiInputBase-input-112 MuiOutlinedInput-input-268 MuiInputBase-inputMultiline-114 MuiOutlinedInput-inputMultiline-270"
          disabled={false}
          name="direction"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows="10"
          value="You need to make something"
        />
      </div>
    </div>
  </div>
  <div>
    <button
      className="MuiButtonBase-root-222 MuiButton-root-234 MuiButton-contained-245 MuiButton-containedPrimary-246 MuiButton-raised-248 MuiButton-raisedPrimary-249 MyForm-button-2"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="Submit"
    >
      <span
        className="MuiButton-label-235"
      >
        Save
      </span>
      <span
        className="MuiTouchRipple-root-275"
      />
    </button>
    <a
      className="MuiButtonBase-root-222 MuiButton-root-234 MuiButton-contained-245 MuiButton-containedSecondary-247 MuiButton-raised-248 MuiButton-raisedSecondary-250 MyForm-button-2"
      href="/recipes"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex="0"
    >
      <span
        className="MuiButton-label-235"
      >
        Cancel
      </span>
      <span
        className="MuiTouchRipple-root-275"
      />
    </a>
  </div>
</form>
`;
